pkg/redirectpolicy/manager.go

```
@@ -0,0 +1,699 @@
+			continue
+		}
+		for _, pod := range pods {
+			if _, ok = pod.namedPorts[namedPort]; ok {
+				// Generate pod backends.
+				for _, ip := range pod.ips {
+					beIP := net.ParseIP(ip)
+					if beIP == nil || bePort.l4Addr.Protocol != feM.feAddr.Protocol {
+						continue
+					}
+					be := backend{
+						IP: net.ParseIP(ip),
+						L4Addr: lb.L4Addr{
+							Protocol: bePort.l4Addr.Protocol,
+							Port:     bePort.l4Addr.Port,
+						},
+					}
+					if feM.feAddr.IP.To4() != nil {
+						if option.Config.EnableIPv4 {
+							bes4 = append(bes4, be)
+						}
+					} else {
+						if option.Config.EnableIPv6 {
+							bes6 = append(bes6, be)
+						}
+					}
+				}
+			}
+			if len(bes4) > 0 {
+				rpm.upsertServiceWithBackends(config, feM, pod.id, bes4)
+			} else if len(bes6) > 0 {
+				rpm.upsertServiceWithBackends(config, feM, pod.id, bes6)
+			}
+		}
+	}
+}
+
+// upsertServiceWithBackends updates policy config internal state and upserts
+// service with the given pod backends.
+func (rpm *Manager) upsertServiceWithBackends(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
+	frontendMapping.backends = backends
+	rpm.policyPods[podID] = append(rpm.policyPods[podID], podPolicyInfo{
+		policyID: config.id,
+		backends: backends,
+	})
+	rpm.upsertService(config, frontendMapping)
+}
+
+// TODO This function along with podMetadata can potentially be removed. We
+// can directly reference the relevant pod metedata on-site.
+func (rpm *Manager) getPodMetadata(pod *slimcorev1.Pod, podIPs []string) *podMetadata {
+	namedPorts := make(serviceStore.PortConfiguration)
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			if port.Name == "" {
+				continue
+			}
+			_, err := u8proto.ParseProtocol(string(port.Protocol))
+			if err != nil {
+				return nil
+			}
+			if port.ContainerPort < 1 || port.ContainerPort > 65535 {
+				return nil
```

Are these supposed to `return nil` or return an error or ignore and continue?

---

Container ports need to be valid, else the redirected traffic will never reach the pod containers.

---

OK, then maybe we should return an error here instead and check it?